standalone/detectors/fasttrack/include/PoolAllocator.h

```diff
@@ -0,0 +1,199 @@
+#ifndef POOL_ALLOCATOR_HEADER_H
+#define POOL_ALLOCATOR_HEADER_H 1
+#pragma once
+
+/*
+ * DRace, a dynamic data race detector
+ *
+ * Copyright 2020 Siemens AG
+ *
+ * Authors:
+ *   Mihai Robescu <mihai-gabriel.robescu@siemens.com>
+ *   Felix Moessbauer <felix.moessbauer@siemens.com>
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include <iostream>
+#include <limits>
+
+#include <ipc/spinlock.h>
+#include "MemoryPool.h"
+
+/**
+ *------------------------------------------------------------------------------
+ *
+ * Header File implementing a pool allocator. It can also be used
+ * in combination with the Segregator class, such as on line 136.
+ *
+ * There is also a thread-safe version. This is currently only
+ * experimental
+ *
+ *------------------------------------------------------------------------------
+ */
+
+template <class T, int threshold, class SmallAllocator, class LargeAllocator>
+class Segregator {
+ public:
+  using value_type = T;
+  using size_type = size_t;
+  using difference_type = ptrdiff_t;
+  using pointer = T*;
+  using const_pointer = const T*;
+  using void_pointer = void*;
+  using const_void_pointer = const void*;
+
+  T* allocate(size_t n) {
+    size_t size = n * sizeof(T);
+    if (size < threshold) {
+      // return new (reinterpret_cast<void*>(SmallAllocator::allocate())) T();
+      return reinterpret_cast<void*>(SmallAllocator::allocate());
+    } else {
+      return reinterpret_cast<void*>(LargeAllocator::allocate());
+    }
+  }
+  void deallocate(T* p, std::size_t n) noexcept {
+    size_t size = n * sizeof(T);
+    if (size < threshold) {
+      return SmallAllocator::deallocate(reinterpret_cast<void*>(p));
+    } else {
+      return LargeAllocator::deallocate(reinterpret_cast<void*>(p));
+    }
+  }
+  template <typename U>
+  struct rebind {
+    using other = Segregator<U, threshold, SmallAllocator, LargeAllocator>;
+  };
+};
+
+template <typename T, size_t numChunks = 512>
+class PoolAllocator {
```

Comment:
Maybe only an interface, as all of them have a static member, that is the memory pool and inheritance would certainly not help code duplication but it will offer a proper framework.

Reply:
@fmoessbauer